### PR TITLE
bpo-36016: Allow gc.getobjects to return the objects in a specific generation

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue36016

### CODE REVIEW

File: Lib/test/test_gc.py [hunk 1]
```diff
@@ -766,6 +766,22 @@ def test_freeze(self):
         gc.unfreeze()
         self.assertEqual(gc.get_freeze_count(), 0)
 
+    def test_get_objects(self):
```
Comment (pganssle) [inline Lib/test/test_gc.py:769]: This test would still pass even if `generation` had no effect, right? I think the important test is to test generations that are being filtered out, so something more like:

```python
self.assertIn(l, gc.get_objects(generation=0))
gc.collect(generation=0)
self.assertNotIn(l, gc.get_objects(generation=0))
```

(Repeat as desired)

Comment (pablogsal) [reply Lib/test/test_gc.py:769]: Done in 604d331

File: Doc/library/gc.rst [hunk 2]
```diff
@@ -63,11 +63,14 @@ The :mod:`gc` module provides the following functions:
...
 
-.. function:: get_objects()
+.. function:: get_objects(generation=None)
 
    Returns a list of all objects tracked by the collector, excluding the list
    returned.
 
+   .. versionchanged:: 3.8
+      If generation is not None, return only the objects tracked by the
+      collector that are in that generation.
```
Comment (methane) [inline Doc/library/gc.rst]: Document about option in function document, and simplify this.
And use ``versionadded`` instead.

File: Modules/gcmodule.c [hunk 3]
```diff
@@ -1502,27 +1502,61 @@ gc_get_referents(PyObject *self, PyObject *args)
 
 /*[clinic input]
 gc.get_objects
+    generation: Py_ssize_t(accept={int, NoneType}) = -1
```
Comment (methane) [inline Modules/gcmodule.c]: Can we hide `-1` from docstring?
How about ``generation: Py_ssize_t(c_default=-1) = None``?

Comment (pablogsal) [reply Modules/gcmodule.c]: Done in 1294e7d

File: Doc/library/gc.rst (hunk 2 above)
Comment (pablogsal) [reply Doc/library/gc.rst]: @methane Done in 1294e7d

I am not sure what do you have in mind for the `versionadded` section. Please, check that what was done in 1294e7d makes sense. :)

File: Modules/gcmodule.c (hunk 3 above)
Comment (pablogsal) [reply Modules/gcmodule.c]: @methane Notice that this hides the `-1` but disallows passing `None` manually:

```python
>>> gc.get_objects(generation=None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'NoneType' object cannot be interpreted as an integer
```

I am not sure if this is better or worse. What do you think?

File: Doc/library/gc.rst [hunk 4]
```diff
@@ -63,11 +63,14 @@ The :mod:`gc` module provides the following functions:
    Return the debugging flags currently set.
 
 
-.. function:: get_objects()
+.. function:: get_objects(generation=None)
 
    Returns a list of all objects tracked by the collector, excluding the list
    returned.
```
Comment (vstinner) [inline Doc/library/gc.rst]: You should document generation parameter, also outside versionchanged note. For example:

```
If *generation* is not None, return only the objects tracked by the
collector that are in that generation.

.. versionchanged:: 3.8
   New *generation* parameter.
```

File: Doc/library/gc.rst (hunk 2 above)
Comment (vstinner) [reply Doc/library/gc.rst]: > And use `versionadded` instead.

IMHO versionadded should be reserved to *new* functions/constants (ex: posix_spawn added to Python 3.8).

File: Doc/whatsnew/3.8.rst [hunk 5]
```diff
@@ -163,6 +163,13 @@ gettext
 Added :func:`~gettext.pgettext` and its variants.
 (Contributed by Franz Glasner, Éric Araujo, and Cheryl Sabella in :issue:`2504`.)
 
+
+gc
+--
+:func:`~gc.get_objects` can now receive an optional parameter indicating a
```
Comment (vstinner) [inline Doc/whatsnew/3.8.rst]:
```suggestion
:func:`~gc.get_objects` can now receive an optional *generation* parameter indicating a
```

File: Modules/gcmodule.c (hunk 3 above)
Comment (vstinner) [reply Modules/gcmodule.c]: >  How about `generation: Py_ssize_t(c_default=-1) = None`?

gc.get_objects(generation=None) must be accepted. Does it work with your proposed syntax?

File: Lib/test/test_gc.py [hunk 6]
```diff
@@ -766,6 +766,30 @@ def test_freeze(self):
...
+        self.assertNotIn(l, gc.get_objects(generation=2))
+        gc.collect(generation=1)
+        self.assertNotIn(l, gc.get_objects(generation=0))
+        self.assertNotIn(l, gc.get_objects(generation=1))
+        self.assertIn(l, gc.get_objects(generation=2))
+        gc.collect(generation=2)
+        self.assertNotIn(l, gc.get_objects(generation=0))
+        self.assertNotIn(l, gc.get_objects(generation=1))
+        self.assertIn(l, gc.get_objects(generation=2))
+
```
Comment (vstinner) [inline Lib/test/test_gc.py]: Please ensure also that gc.get_objects(generation=None) is valid.

Maybe test that gc.get_objects(generation=None) == gc.get_objects()?

**Review by vstinner — CHANGES_REQUESTED** [review]
I like this feature :-)

File: Modules/gcmodule.c (hunk 3 above)
Comment (methane) [reply Modules/gcmodule.c]: `generation: Py_ssize_t(accept={int, NoneType}, c_default="-1") = None`

File: Lib/test/test_gc.py [hunk 7]
```diff
@@ -766,6 +766,30 @@ def test_freeze(self):
...
+        self.assertEqual(gc.collect(), 0)
```
Comment (pitrou) [inline Lib/test/test_gc.py]: I don't think this part of the test is useful. We're testing `gc.get_objects`, not collection of cyclic references.

File: Lib/test/test_gc.py (hunk 6 above)
Comment (pitrou) [reply Lib/test/test_gc.py]: You're going to compare all objects in Python memory for equality?

File: Modules/gcmodule.c [hunk 8]
```diff
@@ -1502,27 +1502,61 @@ gc_get_referents(PyObject *self, PyObject *args)
...
         return NULL;
+    }
+
+    /* If generation is passed, we extract only that generation */
+    if (generation != -1){
+        if (generation >= NUM_GENERATIONS) {
+            PyErr_Format(PyExc_ValueError,
+                         "generation parameter must be less than the number of "
+                         "available generations (%i)",
+                     NUM_GENERATIONS);
```
Comment (pitrou) [inline Modules/gcmodule.c]: Misindent?

File: Lib/test/test_gc.py (hunk 6 above)
Comment (pablogsal) [reply Lib/test/test_gc.py]: I will check only the lenght because checking for equality is going to be super expensive.

**Review by pitrou — COMMENTED** [review]
Looks like a worthwhile addition to me. Some comments below.

File: Modules/gcmodule.c (hunk 3 above)
Comment (pablogsal) [reply Modules/gcmodule.c]: I have use
```
generation: Py_ssize_t(accept={int, NoneType}, c_default="-1") = None`
```
  in 4d8668d

**Comment by pablogsal** [comment]
I have made the requested changes; please review again

**Review by pitrou — APPROVED** [review]
LGTM

File: Doc/whatsnew/3.8.rst [hunk 9]
```diff
@@ -163,6 +163,14 @@ gettext
 Added :func:`~gettext.pgettext` and its variants.
 (Contributed by Franz Glasner, Éric Araujo, and Cheryl Sabella in :issue:`2504`.)
 
+
+gc
+--
```
Comment (vstinner) [inline Doc/whatsnew/3.8.rst:168]: I'm not sure that it's rendered correctly without an empty line between "---" and the new paragraph.

File: Lib/test/test_gc.py [hunk 10]
```diff
@@ -766,6 +766,37 @@ def test_freeze(self):
...
+        self.assertNotIn(l, gc.get_objects(generation=0))
+        self.assertNotIn(l, gc.get_objects(generation=1))
+        self.assertIn(l, gc.get_objects(generation=2))
+        gc.collect(generation=2)
+        self.assertNotIn(l, gc.get_objects(generation=0))
+        self.assertNotIn(l, gc.get_objects(generation=1))
+        self.assertIn(l, gc.get_objects(generation=2))
+        del l
+
+        # Check argument values
```
Comment (vstinner) [inline Lib/test/test_gc.py]: Sometimes adding a comment means that it's time to split your function. Maybe split the function to create "def test_get_objects_arg(self)"?

File: Modules/gcmodule.c [hunk 11]
```diff
@@ -1502,27 +1502,61 @@ gc_get_referents(PyObject *self, PyObject *args)
...
     PyObject* result;
 
     result = PyList_New(0);
-    if (result == NULL)
+    if (result == NULL) {
         return NULL;
+    }
+
+    /* If generation is passed, we extract only that generation */
+    if (generation != -1){
```
Comment (vstinner) [inline Modules/gcmodule.c]:
```suggestion
    if (generation != -1) { 
```

File: Modules/gcmodule.c [hunk 12]
```diff
@@ -1502,27 +1502,61 @@ gc_get_referents(PyObject *self, PyObject *args)
...
+            PyErr_Format(PyExc_ValueError,
+                         "generation parameter must be less than the number of "
+                         "available generations (%i)",
+                          NUM_GENERATIONS);
+            goto error;
+        }
+
+        if (generation < 0) {
+            PyErr_SetString(PyExc_ValueError,
+                            "generation parameter cannot be negative");
```
Comment (vstinner) [inline Modules/gcmodule.c:1538]: Is it "generation argument" or "generation parameter" here? I'm not sure, English is not my first language :-( Same question for the other error message.

File: Modules/gcmodule.c [hunk 13]
```diff
@@ -1502,27 +1502,61 @@ gc_get_referents(PyObject *self, PyObject *args)
...
+        }
+
+        if (append_objects(result, GEN_HEAD(generation))) {
+            goto error;
+        }
+
+        return result;
+    }
+
+    /* If generation is not passed, get all objects from all generations */
```
Comment (vstinner) [inline Modules/gcmodule.c]: nitpick :-)

```suggestion
    /* If generation is not passed or None, get all objects from all generations */
```

File: Modules/gcmodule.c (hunk 12 above)
Comment (pablogsal) [reply Modules/gcmodule.c:1538]: I usually refer to `argument` to refer to something tightly related to the function that needs to be provided while a parameter is something that you can tune or modify to get different behaviour, but is optional.

File: Lib/test/test_gc.py (hunk 10 above)
Comment (pablogsal) [reply Lib/test/test_gc.py]: Done in 5b62014

File: Doc/whatsnew/3.8.rst (hunk 9 above)
Comment (pablogsal) [reply Doc/whatsnew/3.8.rst:168]: Done in 5b62014

**Comment by nascheme** [comment]
I didn't review the code closely but adding this functionality seems like an okay idea.  It is relatively easy to implement and provides a more detailed view into what the GC is doing.

**Pull request merged**